tuning_test.py

```diff
@@ -0,0 +1,266 @@
+
+    @parameterized.named_parameters(
+        dict(
+            testcase_name='reciprocal_10',
+            seed=2,
+            wt_seq=[0, 0, 0, 0],
+            desired_fraction=0.1,
+        ),
+        dict(
+            testcase_name='reciprocal_90',
+            seed=2,
+            wt_seq=[0, 0, 0, 0],
+            desired_fraction=0.9,
+        ),
+        dict(
+            testcase_name='reciprocal_50',
+            seed=2,
+            wt_seq=[0, 0, 0, 0],
+            desired_fraction=0.5,
+        ),
+        dict(
+            testcase_name='reciprocal_66',
+            seed=3,
+            wt_seq=[1, 1, 1, 1],
+            desired_fraction=0.66,
+        ),
+    )
+    def test_tune_epistasis(self, wt_seq, seed, desired_fraction):
+        untuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed)
+        untuned_fraction_adaptive_singles = tuning.get_fraction_adaptive_singles(untuned_landscape)
+
+        tuning_kwargs = tuning.get_tuning_kwargs(
+            untuned_landscape,
+            fraction_reciprocal_adaptive_epistasis=desired_fraction)
+
+        tuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed, **tuning_kwargs)
+        _, actual_fraction = tuning.get_epistasis_stats(tuned_landscape)
+
+        num_singles = len(wt_seq) * (untuned_landscape.vocab_size - 1)
+        num_adaptive_singles = num_singles * untuned_fraction_adaptive_singles
+
+        # This error bound is only exact if all adaptive singles affect unique positions,
+        # otherwise it is more restrictive than necessary.
+        num_adaptive_doubles = comb(num_adaptive_singles, 2)
+        allowed_error = 1.0 / num_adaptive_doubles
+
+        self.assertBetween(actual_fraction,
+                           desired_fraction - allowed_error,
+                           desired_fraction + allowed_error)
+
+    @parameterized.named_parameters(
+        dict(
+            testcase_name='horizon_2',
+            seed=0,
+            wt_seq=[0, 0, 0, 0],
+            desired_horizon=2,
+        ),
+        dict(
+            testcase_name='horizon_5',
+            seed=1,
+            wt_seq=[0, 0, 0, 0],
+            desired_horizon=5,
+        ),
+        dict(
+            testcase_name='horizon_10',
+            seed=2,
+            wt_seq=[0, 0, 0, 0],
+            desired_horizon=10,
+        ),
+        dict(
+            testcase_name='horizon_100',
+            seed=3,
+            wt_seq=[1, 1, 1, 1],
+            desired_horizon=100,
+        ),
+    )
+    def test_tune_epistatic_horizon(self, wt_seq, seed, desired_horizon):
+        untuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed)
+
+        tuning_kwargs = tuning.get_tuning_kwargs(
+            untuned_landscape,
+            epistatic_horizon=desired_horizon)
+
+        tuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed, **tuning_kwargs)
+
+        actual_horizon = tuning.get_epistatic_horizon(tuned_landscape)
```

Comment:
This *should* be ok, but I'm wondering if there's a way to do the direct test - e.g. grab random k-mutants, where k is the epistatic horizon, and show that the properties work out to what you want them to. May not be computationally efficient for long distances - and you'd need to make sure you get good enough sampling to conclude that the horizon is the right length. Not sure if this is worth doing immediately, but something to think about.

Reply:
Yes, I think we can do this. In fact, https://github.com/google-research/slip/pull/8 introduces functions for producing adaptive combinations at desired distances. Our whole epistatic horizon rests on the idea that "on average" the fitness is 0 for adaptive combinations at distance K. One idea is to test that the average of some number of adaptive combos falls within a desired range of 0, but I don't have a good sense for the variance of fitness at distance K, so it's difficult to know what that range should be.

Reply:
I added this as a Todo for now.